Enable WebAPI interface to handle parameters through constructor

phoenix128 opened the pull request:
Add construct parameters injection for WebAPI interfaces.

### Description
WebAPI interfaces were relying on setter/getter methods violating the immutable stateless pattern.
This Pull Request enables the ServiceInputProcessor to get arguments directly from constructor. Setters methods are not required anymore.

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

phoenix128 committed:
Enable WebAPI interface to handle parameters through constructor
DTO setter methods are no more required, parameters can be handled through constructor for immutable state approach
A setter fallback is supported

phoenix128 committed:
FIX for static tests failure
_createFromArray will not be fixed due to its eventual deprecation, since a new immutable approach is provided

phoenix128 committed:
Merge branch '2.3-develop' into webapi-support-immutable-construct-params

phoenix128 committed:
FIX static tests, missing declare(strict_types=1)

phoenix128 committed:
Merge branch 'webapi-support-immutable-construct-params' of github.com:phoenix128/magento2 into webapi-support-immutable-construct-params

VladimirZaets commented on lib/internal/Magento/Framework/Webapi/ServiceInputProcessor.php:
```diff
@@ -82,6 +89,7 @@ public function __construct(
         AttributeValueFactory $attributeValueFactory,
         CustomAttributeTypeLocatorInterface $customAttributeTypeLocator,
         MethodsMap $methodsMap,
+        \Magento\Framework\ObjectManager\ConfigInterface $config,
```
Hi, @phoenix128 , due to [Magento backward-compatible guide](http://devdocs.magento.com/guides/v2.2/contributor-guide/backward-compatible-development/) we can't add required parament to the constructor. 
Please add the optional parameter and take instance from `ObjectManager`, as example:
`$this->storeManager = $storeManager ?: \Magento\Framework\App\ObjectManager::getInstance()>get(\Magento\Store\Model\StoreManagerInterface::class);`

VladimirZaets commented:
Hi @phoenix128  , any updates?

phoenix128 commented:
Hi @VladimirZaets , sorry if I did not answer you before. I've been terribly busy. You are definitely right about that, I must have forgotten it while coding quickly. Fixing in the next commit.

phoenix128 committed:
FIX optional constructor parameter for backward compatiblity

VladimirZaets committed:
Update ServiceInputProcessor.php

Enable WebAPI interface to handle parameters through constructor #14801
- New optional param should be last in the params list.

VladimirZaets committed:
Enable WebAPI interface

VladimirZaets commented:
Hi @phoenix128. Thanks.
Also, optional param should be added to end of the params list. I fixed it.

magento-engcom-team commented:
Hi @phoenix128. Thank you for your contribution. 
Changes from your Pull Request will be available with the upcoming 2.3.0 release.

devamitbera commented:
@phoenix128

Can you please give an example how we can add **Add construct parameters injection for WebAPI interface**
